Fix race condition between gc and committing

csmarchbanks opened the pull request:
Fixes: https://github.com/prometheus/tsdb/issues/230
Perhaps @gouthamve can review?

Signed-off-by: Chris Marchbanks <csmarchbanks@gmail.com>

krasi-georgiev commented:
so we are deleting the series and than on Commit adding these again. 
I am still looking through the code, but can we try to just make the `gc` smarter and maybe keep reference to these until the next commit.
maybe use 
https://github.com/prometheus/tsdb/pull/378/files#diff-0ced3454d447442a96a77b33ebe87300R1198

csmarchbanks commented:
I considered that, but it seemed a little strange to be doing deletes on a Commit, and the gc path would be harder to reason about afterward.

Perhaps another option is to add a `hasPendingSamples` flag to each series when an Add is done, and only delete the series if there are no pending samples? Pending samples would then be cleared on Commit/Rollback.

Let me know what you think about those options/any other options.

krasi-georgiev commented:
> Perhaps another option is to add a hasPendingSamples flag to each series when an Add is done, and only delete the series if there are no pending samples? Pending samples would then be cleared on Commit/Rollback.

that seems like a good solution. Fabian did something similar with the blocks during compaction.

csmarchbanks commented:
Updated, I ended up rearranging some of the error sequencing in `AddFast` in order to only lock the series once. If that is considered backwards breaking (possible that a different error would now be returned), let me know and I will change it.

krasi-georgiev commented:
it should  be fine, 
checking now.

krasi-georgiev commented on head.go:
```diff
@@ -620,21 +620,23 @@ func (a *headAppender) Add(lset labels.Labels, t int64, v float64) (uint64, erro
-	s.Unlock()
-
 	if err != nil {
+		s.Unlock()
```
I bit more simple as:
```
if err := s.appendable(t, v); err != nil {
	s.Unlock()
	return err
}
```

krasi-georgiev commented on head.go:
```diff
@@ -1256,9 +1264,10 @@ type memSeries struct {
+	nextAt     int64 // timestamp at which to cut the next chunk.
+	lastValue  float64
+	sampleBuf  [4]sample
+	addPending bool
```
a little comment of its purpose would be good  and what about if we call it `pendingCommit`

csmarchbanks commented:
Thanks for the feedback! I also like this version much better. I fixed your comments and added another test.

Right now it looks like `Delete` does not have to change, but https://github.com/prometheus/tsdb/pull/270 will need to potentially clear out any pending samples when it is resurrected.

gouthamve commented on head.go:
```diff
@@ -713,6 +715,11 @@ func (a *headAppender) Commit() error {
+		s.series.Lock()
+		s.series.pendingCommit = false
+		s.series.Unlock()
+	}
```
`a.samples` will have different series and here we are acquiring each series' lock. The `s` in `s.series.Lock()` comes from looping.

krasi-georgiev replied:
ooo yeah that is right.

gouthamve commented:
LGTM after @krasi-georgiev's comments are addressed :) Thanks for taking a look at this!

csmarchbanks commented:
Addressed all of @krasi-georgiev comments, thanks again for the review!

krasi-georgiev commented:
Thanks! Job well done!
